infrastructure/instances.tf

```diff
@@ -243,98 +243,265 @@ data "template_file" "nomad_client_script_smasher_smusher" {
 }
 
 ##
-# Autoscaling
+# Autoscaling / Spot Fleets
 ##
 
-resource "aws_launch_configuration" "auto_client_configuration" {
-    # Don't include availability_zones because of:
-    # https://github.com/hashicorp/terraform/issues/15978
-
-    name_prefix = "auto-client-"
-    image_id = "${data.aws_ami.ubuntu.id}"
-    instance_type = "${var.client_instance_type}"
-    security_groups = ["${aws_security_group.data_refinery_worker.id}"]
-    iam_instance_profile = "${aws_iam_instance_profile.data_refinery_instance_profile.name}"
-    depends_on = [
-              "aws_internet_gateway.data_refinery",
-              "aws_instance.nomad_server_1",
-              "aws_ebs_volume.data_refinery_ebs",
-              "aws_instance.pg_bouncer"
-    ]
-    user_data = "${data.template_file.nomad_client_script_smusher.rendered}"
+resource "aws_spot_fleet_request" "cheap_ram" {
+  iam_fleet_role      = "${aws_iam_role.data_refinery_spot_fleet.arn}"
+  allocation_strategy = "diversified"
```

Comment:
Are you sure we want to use `diversified` instead of `lowestPrice`? It seems like lowestPrice would be overall cheaper, but could have more volatility if all our instances are of a single type and then that type's capacity gets snapped up.

However I think that what might be most appropriate for our use case is to follow the directions of the `Configuring Spot Fleet for Cost Optimization and Diversification` section:

> To create a fleet of Spot Instances that is both cheap and diversified, use the lowestPrice allocation strategy in combination with InstancePoolsToUseCount. Spot Fleet automatically deploys the cheapest combination of instance types and Availability Zones based on the current Spot price across the number of Spot pools that you specify. This combination can be used to avoid the most expensive Spot Instances.

What I couldn't determine from the documentation is what happens if there's no spot capacity for some instance types in the `diversified` strategy. Do the spot requests just sit and wait to be fulfilled or do they give up and use what is actually available? It seems to insinuate that at least...

Reply:
My understanding is that `lowestPrice` is for riding the variance in spot price for a single instance type across regions and AZs. Since we're only operating on a single AZ and our limiting factor is capacity, I think that `diversified` is the correct strategy here.